File: ChatApp/frontend/application/dto/ws-messages.ts
export interface WsToolCall {
  type: "tool_call";
  tool: string;
  input: Record<string, unknown>;
}

export interface WsToolResult {
  type: "tool_result";
  tool: string;
  result: string;
}

export interface WsStream {
  type: "stream";
  content: string;
}

export interface WsReply {
  type: "reply";
  content: string;
}

export interface WsError {
  type: "error";
  message: string;
}

export interface WsStatus {
  type: "status";
  content: string;
}

export interface WsCleared {
  type: "cleared";
}

export type WsServerMessage =
  | WsToolCall
  | WsToolResult
  | WsStream
  | WsReply
  | WsError
  | WsStatus
  | WsCleared;

File: ChatApp/backend/application/dto/ws_messages.py
from typing import Literal

from pydantic import BaseModel


class WsIncomingMessage(BaseModel):
    type: str
    content: str = ""


class WsToolCall(BaseModel):
    type: Literal["tool_call"] = "tool_call"
    tool: str
    input: dict


class WsToolResult(BaseModel):
    type: Literal["tool_result"] = "tool_result"
    tool: str
    result: str


class WsStream(BaseModel):
    type: Literal["stream"] = "stream"
    content: str


class WsReply(BaseModel):
    type: Literal["reply"] = "reply"
    content: str


class WsError(BaseModel):
    type: Literal["error"] = "error"
    message: str


class WsStatus(BaseModel):
    type: Literal["status"] = "status"
    content: str


class WsCleared(BaseModel):
    type: Literal["cleared"] = "cleared"
